Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp

```diff
@@ -171,63 +468,109 @@ void Client::LookForWorkers()
         return;
     }
 
-    // randomize the start index to better distribute workers when there
+    // find someone to connect to
+    // randomize the start indices to better distribute workers when there
     // are many workers/clients - otherwise all clients will attempt to connect
     // to the same subset of workers
-    Random r;
-    size_t startIndex = r.GetRandIndex( (uint32_t)numWorkers );
-
-    // find someone to connect to
-    for ( size_t j=0; j<numWorkers; j++ )
+    // get a random worker
+    Random randomWorker;
+    size_t workerStartIndex = randomWorker.GetRandIndex( (uint32_t)numWorkers );
+    Array< AString > newlyExcludedWorkers;
+    for ( size_t i=0; i<numWorkers; ++i )
     {
-        const size_t i( ( j + startIndex ) % numWorkers );
+        const size_t j( ( i + workerStartIndex ) % numWorkers );
+        const AString & worker = m_WorkerList[ j ];
+        ServerState * ss = nullptr;
+        for ( size_t m=0; m<numWorkers; ++m )
+        {
+            ServerState * const stateToCheck = m_ServerList[ m ];
+            if ( stateToCheck && stateToCheck->m_RemoteName == worker )
+            {
+                ss = stateToCheck;
+                break;
+            }
+        }
 
-        ServerState & ss = m_ServerList[ i ];
-        if ( ss.m_Connection )
+        if ( ss == nullptr || ss->m_Connection )
         {
             continue;
         }
 
+        MutexHolder mhSS( ss->m_Mutex );
+
         // ignore blacklisted workers
-        if ( ss.m_Blacklisted )
+        if ( ss->m_Blacklisted )
         {
+            ExcludeWorker( *ss, newlyExcludedWorkers );
             continue;
         }
 
-        // lock the server state
-        MutexHolder mhSS( ss.m_Mutex );
-
-        ASSERT( ss.m_Jobs.IsEmpty() );
+        ASSERT( ss->m_Jobs.IsEmpty() );
 
-        if ( ss.m_DelayTimer.GetElapsed() < CONNECTION_REATTEMPT_DELAY_TIME )
+        float timeSinceLastRetry = ss->m_ConnectionDelayTimer.GetElapsed();
+        if ( timeSinceLastRetry <= ss->m_ConnectionRetryIntervalSec )
         {
             continue;
         }
 
-        DIST_INFO( "Connecting to: %s\n", m_WorkerList[ i ].Get() );
-        const ConnectionInfo * ci = Connect( m_WorkerList[ i ], m_Port, 2000, &ss ); // 2000ms connection timeout
+        ss->m_ConnectionDelayTimer.Start();  // reset time
```

Comment:
We now reset the connection timer even if we successfully connect. Is that needed?